src/lsp4clj/coercer.clj

```diff
@@ -0,0 +1,432 @@
+
+(s/def :document-symbol/selection-range ::range)
+
+(s/def :document-symbol/detail string?)
+
+(s/def ::document-symbol (s/keys :req-un [::name
+                                          :symbol/kind
+                                          ::range
+                                          :document-symbol/selection-range]
+                                 :opt-un [:document-symbol/detail :document-symbol/children]))
+
+(s/def :document-symbol/children (s/coll-of ::document-symbol))
+
+(s/def ::document-symbols (s/coll-of ::document-symbol))
+
+(s/def ::document-highlight (s/keys :req-un [::range]))
+
+(s/def ::document-highlights (s/coll-of ::document-highlight))
+
+(s/def ::symbol-information (s/keys :req-un [::name :symbol/kind ::location]))
+
+(s/def ::workspace-symbols (s/coll-of ::symbol-information))
+
+(s/def ::severity integer?)
+
+(s/def ::code (s/conformer name))
+
+(s/def ::diagnostic (s/keys :req-un [::range ::message]
+                            :opt-un [::severity ::code ::tag ::source ::message]))
+(s/def ::diagnostics (s/coll-of ::diagnostic))
+(s/def ::publish-diagnostics-params (s/keys :req-un [::uri ::diagnostics]))
+
+(s/def ::marked-string (s/and (s/or :string string?
+                                    :marked-string (s/keys :req-un [::language ::value]))
+                              (s/conformer second)))
+
+(s/def :markup/kind #{"plaintext" "markdown"})
+(s/def :markup/value string?)
+(s/def ::markup-content (s/keys :req-un [:markup/kind :markup/value]))
+
+(s/def ::contents (s/and (s/or :marked-strings (s/coll-of ::marked-string)
+                               :markup-content ::markup-content)
+                         (s/conformer second)))
+
+(s/def ::hover (s/keys :req-un [::contents]
+                       :opt-un [::range]))
+
+(s/def :command/title string?)
+(s/def :command/command string?)
+(s/def :command/arguments (s/coll-of any?))
+
+(s/def ::command (s/keys :req-un [:command/title :command/command]
+                         :opt-un [:command/arguments]))
+
+(def show-message-type-enum
+  {:error 1
+   :warning 2
+   :info 3
+   :log 4})
+
+(s/def :show-message/type (s/and keyword?
+                                    show-message-type-enum
+                                    (s/conformer show-message-type-enum)))
+
+(s/def :show-message/message string?)
+
+(s/def ::show-message (s/keys :req-un [:show-message/type
+                                       :show-message/message]))
+
+(s/def :show-message-request-action/title string?)
+
+(s/def :show-message-request/action (s/keys :req-un [:show-message-request-action/title]))
+
+(s/def :show-message-request/actions (s/coll-of :show-message-request/action))
+
+(s/def ::show-message-request (s/keys :req-un [:show-message/type
+                                               :show-message/message]
+                                      :opt-un [:show-message-request/actions]))
+
+;; (def work-done-progress-kind-enum
```

Comment:
why commented?

Reply:
Progress notifications were one of the first things I implemented. Instead of doing conformation, I made `lsp4clj.json-rpc.messages/work-done-progress`, a function which does the same manipulation as this coercer code did—converting 0 to `:begin`, 100 to `:done` and other related transformations. I thought I would do that for all the output, replacing all the conformation. But that ended up being too much work.

So, this piece of the coercer got commented out, while other pieces got left in, although modified to remove the Java conversions. I'd be happy to uncomment it, if you'd like, but there won't be any users of it.

Reply:
I think we can remove it so?